server/plugins/BelaUGens.cpp

```diff
@@ -0,0 +1,1157 @@
+        // calling [], you may have been passed a reference to last instead of
+        // a pointer into the buffer itself. Here, we ensure we put it back into the buffer
+        if (lastTouched)
+            buffer[count - 1] = last;
+    }
+    float& operator[](unsigned int n) {
+        if (n > count)
+            lastTouched = true;
+        return at(n);
+    }
+
+private:
+    bool lastTouched;
+};
+
+class AccessBufferReader : public AccessBuffer {
+public:
+    AccessBufferReader(float* buffer, unsigned int count): AccessBuffer(buffer, count) {};
+    const float& operator[](unsigned int n) const { return at(n); }
+};
+
+static InterfaceTable* ft;
+
+static inline void BelaUgen_init_output(Unit* unit) { (unit->mCalcFunc)(unit, 1); }
+
+static inline void BelaUgen_disable(Unit* unit) {
+    SETCALC(ClearUnitOutputs);
+    BelaUgen_init_output(unit);
+}
+
+struct MultiplexAnalogIn : public Unit {};
+
+
+struct AnalogIn : public Unit {
+    int mAnalogPin;
+};
+
+struct AnalogOut : public Unit {
+    int mAnalogPin;
+};
+
+// static digital pin, static function (in)
+struct DigitalIn : public Unit {
+    int mDigitalPin;
+};
+
+// static digital pin, static function (out) - uses DigitalWrite and a check whether value changed
+struct DigitalOut : public Unit {
+    int mDigitalPin;
+    int mLastOut;
+};
+
+// flexible digital pin, flexible function (in or out)
+struct DigitalIO : public Unit {
+    int mDigitalPin;
+    int mLastDigitalIn;
+    int mLastDigitalOut;
+};
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+
+void MultiplexAnalogIn_next_aaa(MultiplexAnalogIn* unit, int inNumSamples) {
+    World* world = unit->mWorld;
+    BelaContext* context = world->mBelaContext;
+
+    float* fin = IN(0); // analog in pin, can be modulated
+    float* fmux = IN(1); // mux channel, can be modulated
+    float* out = ZOUT(0);
+    int analogPin = 0;
+    int muxChannel = 0;
+    float analogValue = 0;
+
+    // context->audioFrames should be equal to inNumSamples
+    //   for(unsigned int n = 0; n < context->audioFrames; n++) {
+    for (unsigned int n = 0; n < inNumSamples; n++) {
+        analogPin = (int)fin[n];
+        muxChannel = (int)fmux[n];
+        if ((analogPin < 0) || (analogPin >= context->analogInChannels) || (muxChannel < 0)
+            || (muxChannel > context->multiplexerChannels)) {
+            rt_fprintf(stderr, "MultiplexAnalogIn warning: analog pin must be between %i and %i, it is %i \n", 0,
+                       context->analogInChannels, analogPin);
+            rt_fprintf(stderr, "MultiplexAnalogIn warning: muxChannel must be between %i and %i, it is %i \n", 0,
+                       context->multiplexerChannels, muxChannel);
+        } else {
+            analogValue = multiplexerAnalogRead(
+                context, analogPin, muxChannel); // is there something like NI? analogReadNI(context, 0, analogPin);
+        }
+        *++out = analogValue;
```

Comment:
I don't think this particular ugen has ever been tested properly. I guess this assumed that `ZOUT` had a `ZOFF` of 1?

Reply:
ahh i see now, it is correct after all, thanks! but that brings up 2 other points:
1. i think the Unroll.h macros should generally all be used together to be as clear as possible and to make sure switching between preincrement/postincrement doesn't break code.
2. has this code (or similar code) been benchmarked on Bela to see whether preincrement is actually faster on the ARM chip used on the Bela board? it's not necessary that you or anyone does this, but it would be good to know. the only benchmark blog posts noted in the code (http://mcld.co.uk/blog/2008/efficiency-geek-zeroing-data-in-cc-optimisation.html, http://mcld.co.uk/blog/2008/efficiency-geek-2-copying-data-in-cc-optimisation.html) were done on PowerPC and x86_64.

in any case i will change these `*++out` to `ZXP(out)` to be more consistent.

Reply:
> 1. i think the Unroll.h macros should generally all be used together to be as clear as possible and to make sure switching between preincrement/postincrement doesn't break code.

Hard agree. Again, I think this specific UGen here was not quite finished.

>  Bela to see whether preincrement is actually faster on the ARM chip used on the Bela board?

I don't think so. However, I'd hope that in 2021 we can trust compilers know better than us ...

> in any case i will change these *++out to ZXP(out) to be more consistent.

For the reason above, I'd suggest to avoid `ZOUT()` whenever possible for new code and only use `OUT()` to avoid further confusion ...

Reply:
> However, I'd hope that in 2021 we can trust compilers know better than us ...

There is a difference in the generated assembly for a similar loop -- https://godbolt.org/z/jT4qcn -- but I don't know of a site online where you can benchmark code on an ARM processor or I'd check to see if there's an actual speed difference. In general though I'm with you there, most of the time the compiler knows best. :)

> For the reason above, I'd suggest to avoid ZOUT() whenever possible for new code and only use OUT() to avoid further confusion ...

i'm also fine with that! i'll do it for these UGens since i'm pretty sure in this case the digital or analog pin IO will dominate execution time anyway. personally though, as long as the macro switching leads to a difference in generated assembly, i would want to check on x86 and ARM whether there is a noticeable difference in execution time before totally retiring them.

Reply:
> There is a difference in the generated assembly for a similar loop

I'm not familiar with ARM assembly, but the difference doesn't like significant to me. For x86-64, clang produces *identical* assembly. I don't see why this shouldn't be the case for other targets, as the code is 100% semantically equivalent.

Questions like pre-increment VS post-increment or for-loop VS while-loop may have been a thing in the 90s, but in 2020 they are completely irrelevant.

Generally, the macros in `Unroll.h` are legacy and should be avoided in new code! They only make the code more difficult to understand. Just use a plain for loop and let the compiler (and auto-vectorizer) do its job :-)

Reply:
there was also a difference in generated GCC assembly when I checked, too. I'm not comfortable dismissing this as unimportant without statistics to back that up, these are quite literally the critical loops of supercollider.

Reply:
@brianlheim I checked the GCC x86_64 assembly. There is minor difference (version A adds an offset to the frame pointer at the beginning of the loop body and version B subtracts an offest from the frame pointer at the end of the loop body), but the result is the same.

Again, stuff like `FOR_IS_FASTER`, `WHILE_IS_FASTER`, `PREINCREMENT_IS_FASTER` and `POSTINCREMENT_IS_FASTER` is entirely obsolete in 2020. The optimizer performs a semantic analysis of the code. It doesn't matter if you use a for-loop, a while-loop, C++ range based for-loop, std::for_each::, etc. - they all *mean* the same thing.

FWIW, https://doc.sccode.org/Reference/ServerPluginAPI.html also discourages the use of the `Unroll.h` macros:

> The macros in this section are legacy features. They are seen in many of SuperCollider's built-in ugens, and are intended to provide more efficient alternatives to the standard for (int i = 0; i < inNumSamples; i++) { out[i] = in[i] } loop. These efficiency savings are negligible on modern systems and use of these macros is not recommended, especially since they make debugging difficult.

(I would go further and claim that for modern compilers there are no efficiency savings at all.)

---

However, there are certainly situations where the compiler needs your help. Often it has to pessimistically assume that operations might have side effects and therefore can't auto-vectorize. For example, it might assume that writing a sample to a UGen output might change a UGen input for the next sample, but you as a developer know that this is never the case. Sometimes things like these can be solved with the (non-standard) `restrict` keyword, other times you need to manually unroll the loop.

Reply:
sorry, but i have been bitten too many times by almost-identical assembly having surprises that i don't trust my intuition on this anymore. plus, my low-latency experience is not with audio programming nor did it involve many tight loops, so again my intuitions/rules of thumb aren't as applicable. it's also unclear whether "modern systems" is meant to apply to ARM. i will do some quick benchmarks myself and then probably agree with you. :)

Reply:
> it's also unclear whether "modern systems" is meant to apply to ARM

I would rather say "modern compilers". The actual target is not relevant, since a for-loop is a general concept that compilers understand. That's why they are able to hoist computations out of the loop body, do loop unrolling, auto-vectorize, replace whole loops with a library function call, etc.

> i will do some quick benchmarks myself and then probably agree with you. :)

Go ahead :-) In the meantime, I'll learn some basic ARM assembly, so I can *prove* that the generated code is equivalent (just like I can for the x86_64 output).

Reply:
Here are 8 different ways how to clear a float array, including the obscure pre-increment style: https://godbolt.org/z/q3GvMG

Note how GCC and clang just produce a call to `memset` for all versions? (Clang even produces identical assembly, while GCC has some slight variations in the function call setup.) That's what I meant when I said that loops are a general concept that is understood by compilers. We simply don't have to worry (anymore) how to write a loop, the compiler will recognize it anyway, so we can just pick a style that is easy to read :-)

Reply:
@Spacechild1 i am pretty familiar with the concepts we are talking about, thanks, i just want to investigate this isolated case myself.

Reply:
ok, just did a test on my mac and rpi and the differences in timing are very close compared to the benchmark time variance, so i'm convinced!

@Spacechild1 for some background -- i used to collaborate with someone who gave a lot of advice about codegen and performance. most were right, but once in a while i would dig deeper and benchmark and find that some actually no longer applied, or didn't apply in the specific cases i was looking at. this person was just as certain as you are of their explanations, and actually provided similarly convincing reasoning. this is not to say that i don't trust you or find your justifications adequate, and i'm sorry if it came across that way! -- on the contrary i'm inclined to agree with you. it's just that if i want to make this recommendation to others, i'd rather check myself on a real system. a bit paranoid, i know, but this also helps me to build my own intuition. i do appreciate you taking the time, and these other examples of optimisation are also compelling.

there is a third reason now to drop these macros, besides the ones already mentioned -- clang-format is not a fan of [LOOP](https://github.com/supercollider/supercollider/blob/develop/server/plugins/OscUGens.cpp#L1099) and [LOOP1](https://github.com/supercollider/supercollider/blob/develop/server/plugins/OscUGens.cpp#L1257-L1259). it's nice that we can fix that now just by removing the macros, instead of some other fiddly way like adding clang-format pragmas.

Reply:
I'm glad your benchmarks confirm my predictions, otherwise I would have questioned my sanity :-) You're of course right not to take anything for granted and benchmark yourself. In that specific case I was just so absolutely certain that I wanted to spare you the work. (But I was a bit wrong, see below.)

---

In the meantime, I've figured out the ARM assembly syntax in your example:

Left: `ldr r1, [r6, #4]!` - pre-index with write back (= get memory address in `r6`, add 4, write back to `r6` and store the memory content in `r1`) - corresponds to `*(++ptr)`
Right: `r1, [r6], #4` - post-index with write back (= get memory address in `r6`, store the memory content in `r1`, add 4 to address and write back to `r6`) - corresponds to `*(ptr++)`

See section "4.6.3.1 Operand Syntaxes for Instructions" in http://downloads.ti.com/docs/esd/SPNU118O/Content/SPNU118O_HTML/assembler_description.html.

Here's a nice tutorial: https://azeria-labs.com/memory-instructions-load-and-store-part-4/

---

Actually, I was asking myself why your example would't produce *identical* assembly and then I realized that you forgot the initial `out -= ZOFF; bounds -= ZOFF` statements. Here's the results for the corrected code:

The GCC ARM compiler adds two initial `sub` statements for the pre-increment version, but the loop body is equivalent: https://godbolt.org/z/97Eq8b.

Clang ARM is a bit more clever and produces almost identical output: https://godbolt.org/z/boPfqa. (I don't know why the left version uses a seperate `add r6, r6, #4` statement instead of the post-index store `r0, [r6], #4` in the right version...)

GCC x86_64 uses equivalent code in the loop body, but the pre-increment version has some very minor overhead in the loop setup: https://godbolt.org/z/4rs8qG

Finally, clang x86_64 produces *identical* assembly: https://godbolt.org/z/sf6MT9

This means that the weird obfuscated pre-increment version actually manages to confuse several compilers and make them produce sub-optimal assembly (although the overhead shouldn't be significant). So your scepticism was warranted! However, the results still support my initial advice regarding loops: write what you *intend* and don't try to outsmart the compiler - it might backfire! Or put in another way: just use a bog-standard for-loop :-)